### ISSUE TITLE
Confusing borrow check error when function argument doesn't match type signature

### ISSUE THREAD

**Issue by tmandry** [rust-lang/rust#75929]
```rust
struct Db<'a>(&'a str);
impl<'a> Db<'a> {
    fn new(x: &'a str) -> Self {
        Self(x)
    }
}

fn foo<'a>(buf: &'a mut str) -> Db<'a> {
    Db::new(&buf)
}
```

gives

```
error[E0515]: cannot return value referencing function parameter `buf`
 --> src/lib.rs:9:5
  |
9 |     Db::new(&buf)
  |     ^^^^^^^^----^
  |     |       |
  |     |       `buf` is borrowed here
  |     returns a value referencing data owned by the current function
```

when all that's needed is a reborrow: `&*buf`.

In this case `&buf` is `&&mut str` which doesn't match the type signature of `Db::new`, so it doesn't seem like it should have gotten past type check.

**Comment by SNCPlay42** [comment]
> In this case `&buf` is `&&mut str` which doesn't match the type signature of `Db::new`, so it doesn't seem like it should have gotten past type check.

That's just deref coercion (and `&mut T => &T` coercion), right?

e.g. this compiles:

```rust
fn x(y:&i32){}

fn main() {
    x(&&&mut&mut&mut 1);
}
```

**Comment by tmandry** [comment]
> That's just deref coercion (and `&mut T => &T` coercion), right?

Yes! I came back here to say this. Edited the description.

**Comment by nbdd0121** [comment]
I didn't see anything wrong with the diagnostics, as this is indeed a lifetime issue. `&&mut T` could be coerced to `&T`, so this passes typeck. However, obviously `buf` is a local variable so `&buf` has its lifetime restricted to the scope of this function. Region inference will cause the type of that expression to be `Db<'buf>`, so it triggers this error.

If you spell out lifetime of Db, it will gives you a different error.
```rust
struct Db<'a>(&'a str);
impl<'a> Db<'a> {
    fn new(x: &'a str) -> Self {
        Self(x)
    }
}

fn foo<'a>(buf: &'a mut str) -> Db<'a> {
    Db::<'a>::new(&buf)
}
```
gives
```
error[E0597]: `buf` does not live long enough
  --> src/lib.rs:9:19
   |
8  | fn foo<'a>(buf: &'a mut str) -> Db<'a> {
   |        -- lifetime `'a` defined here
9  |     Db::<'a>::new(&buf)
   |     --------------^^^^-
   |     |             |
   |     |             borrowed value does not live long enough
   |     argument requires that `buf` is borrowed for `'a`
10 | }
   | - `buf` dropped here while still borrowed
```

**Comment by tmandry** [comment]
The error may be correct, but it's confusing and doesn't offer any guidance, particularly for beginners who might not realize that they can pass a borrow "directly" without needing an &. What I think should happen here is a suggestion to do just that.

**Comment by camelid** [comment]
@RodBurman the "new diagnostic" you posted is the same as the one in the original issue, is it not?

**Comment by RodBurman** [comment]
@camelid yes it is - I managed to confuse myself and then spread the contagion :(. I have deleted the offending comment.
 What I should really have said was, problem remains in stable release 1.97.1

**Comment by camelid** [comment]
No worries, thanks for clearing it up.
